include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,405 @@
+class mem_fun_ptr<T C::*> {
+  T C::*_pm;
+
+private:
+  // `(t1.*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a class
+  // `C` and `is_base_of_v<T, remove_reference_t<decltype(t1)>>` is `true`;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R _call(ref_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept((std::forward<T1>(t1).*pm)(std::forward<Tn>(tn)...))) {
+    return (std::forward<T1>(t1).*pm)(std::forward<Tn>(tn)...);
+  }
+
+  // `(t1.get().*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a
+  // class `C` and `remove_cvref_t<decltype(t1)>` is a specialization of
+  // `reference_wrapper`;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R
+  _call(ref_wrapper_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept((t1.get().*pm)(std::forward<Tn>(tn)...))) {
+    return (t1.get().*pm)(std::forward<Tn>(tn)...);
+  }
+
+  // `((*t1).*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a
+  // class `C` and `t1` does not satisfy the previous two items;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R _call(ptr_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept(((*std::forward<T1>(t1)).*pm)(std::forward<Tn>(tn)...))) {
+    return ((*std::forward<T1>(t1)).*pm)(std::forward<Tn>(tn)...);
+  }
+
+public:
+  constexpr mem_fun_ptr(T C::*pm) noexcept : _pm(pm) {}
+
+  template <typename T1,
+            typename... Tn,
+            typename PT1 = typename p0704<T, T1>::type,
+            typename Tag = typename dispatch_mem_ptr<C, PT1>::type,
+            typename R =
+                typename mem_fun_ptr_result<Tag, T C::*(PT1&&, Tn&&...)>::type,
+            typename = typename apply_clang_workaround<C, T, T1>::type>
+
+  constexpr R operator()(T1&& t1, Tn&&... tn) const noexcept(noexcept(
+      _call<R>(Tag{}, _pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...))) {
+    return _call<R>(Tag{}, _pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...);
+  }
+};
+
+} // namespace _invoke
+
+// We don't use `std::is_function<T>` here since some libstdc++ versions fail to
+// classify function types with noexcept specifiers.
+
+template <typename T>
+struct is_function {
+  static constexpr bool value =
+      !std::is_const<T const>::value && !std::is_reference<T>::value;
+};
+
+template <typename T, typename Enable = void>
+struct mem_fn_result {};
+
+template <typename T, typename C>
+struct mem_fn_result<T C::*,
+                     typename std::enable_if<!is_function<T>::value>::type> {
+  using type = _invoke::mem_obj_ptr<T C::*>;
+};
+
+template <typename T, typename C>
+struct mem_fn_result<T C::*,
+                     typename std::enable_if<is_function<T>::value>::type> {
+  using type = _invoke::mem_fun_ptr<T C::*>;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+namespace _invoke {
+
+template <typename F, typename Enable = void>
+struct dispatch_invoke;
```

Comment:
Seems like we can use `std::conditional` here as well.
```
template <typename F>
struct dispatch_invoke
    : std::conditional<
          std::is_member_pointer<
              typename std::remove_reference<F>::type>::value,
          typename mem_fn_result<typename std::remove_cv<
              typename std::remove_reference<F>::type>::type>::type,
          F&&> {};
```

Reply:
This seems to be instantiating `mem_fn_result<...>::type` for non-member-pointers, which doesn't exist and would cause a hard error.

Reply:
What we could do here is use traditional substitution, and partially specialize for `T C::*`.

Reply:
Except that the member pointer might be cv-ref qualified, so we would need to decay `F` first.